## Import Module

In [2]:
# Data Analysis
import pandas as pd
import numpy as np

# Data Visualization
import seaborn as sn

## Case 1

### Informasi Dataset ride_sharing_new.csv

In [3]:
ride_sharing_new = pd.read_csv('ride_sharing_new.csv')
ride_sharing_new.head(10)

duration  station_A_id  \
0  12 minutes            81   
1  24 minutes             3   
2   8 minutes            67   
3   4 minutes            16   
4  11 minutes            22   
5  10 minutes            22   
6  16 minutes            67   
7   9 minutes            30   
8  21 minutes            81   
9   5 minutes            30   

                                      station_A_name  station_B_id  \
0                                 Berry St at 4th St           323   
1       Powell St BART Station (Market St at 4th St)           118   
2  San Francisco Caltrain Station 2  (Townsend St...            23   
3                            Steuart St at Market St            28   
4                              Howard St at Beale St           350   
5                              Howard St at Beale St             6   
6  San Francisco Caltrain Station 2  (Townsend St...           107   
7     San Francisco Caltrain (Townsend St at 4th St)            97   
8                                 Berry St at 4th St            16   
9     San Francisco Caltrain (Townsend St at 4th St)            62   

                    station_B_name  bike_id  user_type  user_birth_year  \
0               Broadway at Kearny   5480.0          2             1959   
1  Eureka Valley Recreation Center   5193.0          2             1965   
2    The Embarcadero at Steuart St   3652.0          3             1993   
3     The Embarcadero at Bryant St   1883.0          1             1979   
4             8th St at Brannan St   4626.0          2             1994   
5    The Embarcadero at Sansome St   3279.0          2             1979   
6            17th St at Dolores St   1035.0          2             1981   
7            14th St at Mission St   5286.0          2             1991   
8          Steuart St at Market St      NaN          1             1982   
9      Victoria Manalo Draves Park    333.0          1             1994   

  user_gender  
0        Male  
1        Male  
2        Male  
3        Male  
4        Male  
5        Male  
6        Male  
7        Male  
8      Female  
9        Male

In [4]:
ride_sharing_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25760 entries, 0 to 25759
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   duration         25760 non-null  object 
 1   station_A_id     25760 non-null  int64  
 2   station_A_name   25760 non-null  object 
 3   station_B_id     25760 non-null  int64  
 4   station_B_name   25760 non-null  object 
 5   bike_id          25755 non-null  float64
 6   user_type        25760 non-null  int64  
 7   user_birth_year  25760 non-null  int64  
 8   user_gender      25760 non-null  object 
dtypes: float64(1), int64(4), object(4)
memory usage: 1.8+ MB


In [5]:
# looking for missing value
missing_values = pd.isnull(ride_sharing_new).sum()
missing_values

duration           0
station_A_id       0
station_A_name     0
station_B_id       0
station_B_name     0
bike_id            5
user_type          0
user_birth_year    0
user_gender        0
dtype: int64

Sebelum dilakukan cleaning data, sebaiknya kita mengetahui permasalahan apa yang terjadi pada kolom tersebut sehingga kita dapat menentukan penyelesaian apa yang sesuai dengan permasalahan missing value. <br>
Beberapa cara yang dapat dilakukan dalam mengatasi missing value, diantaranya :
1. Menghilangkan atau menghapus data yang hilang. Cara ini banyak digunakan apabila data yang kosong terlalu banyak lebih dari 60%.
2. Melakukan inputting value dengan data yang menyerupai seperti minimum value, maximum value, mean, median, atau modus. Penggunaan mean banyak digunakan apabila data tersebut adalah data kontinue, bukan kategorik data. Sedangkan modus digunakan ketika kolom yang kosong berisi informasi berupa data kategorik.
3. Melakukan manual inputting value dengan data yang sebenarnya (cara ini membutuhkan high effort untuk mencari data yang hilang tersebut)
4. Menentukan custom value berdasarkan asumsi bisnis yang terjadi.


In [6]:
# Rasio Missing Values
ratio_missing_value = missing_values.div(len(ride_sharing_new))
ratio_missing_value

duration           0.000000
station_A_id       0.000000
station_A_name     0.000000
station_B_id       0.000000
station_B_name     0.000000
bike_id            0.000194
user_type          0.000000
user_birth_year    0.000000
user_gender        0.000000
dtype: float64

<b>Tidak ada Drop Data</b>, karena Rasio missing values masih di bawah 60%, <br>
<u>Missing value terdapat pada bike_id</u>
<br>
<br>
<p style="color:green"> Keseluruhan data termasuk bersih.

### Summary Dataset

In [7]:
ride_sharing_new_desc = ride_sharing_new.describe()

# Function to format the floating-point numbers
def format_floats(x):
    return f'{x:.6f}'.rstrip('0').rstrip('.') if isinstance(x, (float, np.float64)) else x

# Apply the formatting function to the DataFrame
formatted_desc = ride_sharing_new_desc.applymap(format_floats)

# Apply the background gradient
styled_desc = formatted_desc.style.background_gradient(cmap='mako')

# Display the styled DataFrame
styled_desc

C:\Users\lenovo\AppData\Local\Temp\ipykernel_22344\605183901.py:8: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  formatted_desc = ride_sharing_new_desc.applymap(format_floats)


## case 2

### Apakah terdapat missing value

In [8]:
ride_sharing_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25760 entries, 0 to 25759
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   duration         25760 non-null  object 
 1   station_A_id     25760 non-null  int64  
 2   station_A_name   25760 non-null  object 
 3   station_B_id     25760 non-null  int64  
 4   station_B_name   25760 non-null  object 
 5   bike_id          25755 non-null  float64
 6   user_type        25760 non-null  int64  
 7   user_birth_year  25760 non-null  int64  
 8   user_gender      25760 non-null  object 
dtypes: float64(1), int64(4), object(4)
memory usage: 1.8+ MB


In [9]:
missing_values

duration           0
station_A_id       0
station_A_name     0
station_B_id       0
station_B_name     0
bike_id            5
user_type          0
user_birth_year    0
user_gender        0
dtype: int64

In [10]:
ratio_missing_value

duration           0.000000
station_A_id       0.000000
station_A_name     0.000000
station_B_id       0.000000
station_B_name     0.000000
bike_id            0.000194
user_type          0.000000
user_birth_year    0.000000
user_gender        0.000000
dtype: float64

Ya, terdapat missing value pada field <u>bike_id</u>

### Handling missing value tersebut

#### Handling Missing Value

In [11]:
ride_sharing_new['bike_id'].describe() #describe bike_id karena memiliki dType float64

count    25755.000000
mean      4108.053388
std       1576.015656
min         11.000000
25%       3108.000000
50%       4821.000000
75%       5258.000000
max       6638.000000
Name: bike_id, dtype: float64

tidak terdapat Nan atau missing value pada kolom bike_id<br>
mean = 4108<br>
median = 4821
<br>
perbedaan yang tidak terlalu jauh ini dapat diatasi dengan metode missing value yaitu input nilai missing value dengan mean

In [12]:
ride_sharing_new['bike_id'].fillna(int(ride_sharing_new['bike_id'].mean()), inplace=True) # mengubah Nan dengan nilai mean

pd.isnull(ride_sharing_new).sum() #cek apakah masih terdapat missing value

duration           0
station_A_id       0
station_A_name     0
station_B_id       0
station_B_name     0
bike_id            0
user_type          0
user_birth_year    0
user_gender        0
dtype: int64

## Case 3

### Data Duplicate

In [13]:
# Mengidentifikasi apakah terdapat duplicate value pada tabel dengan acuan kolom PassengerId
ride_sharing_new[ride_sharing_new["duration"].duplicated(keep=False)].sort_values("duration")   

duration  station_A_id  \
3780    1 minutes            16   
21077   1 minutes            22   
9407    1 minutes            67   
16632   1 minutes            21   
6659    1 minutes            30   
...           ...           ...   
17008  93 minutes            15   
14841  93 minutes            30   
25079  93 minutes             3   
25078  99 minutes             3   
11420  99 minutes            15   

                                          station_A_name  station_B_id  \
3780                             Steuart St at Market St            16   
21077                              Howard St at Beale St            25   
9407   San Francisco Caltrain Station 2  (Townsend St...            67   
16632   Montgomery St BART Station (Market St at 2nd St)            21   
6659      San Francisco Caltrain (Townsend St at 4th St)            80   
...                                                  ...           ...   
17008  San Francisco Ferry Building (Harry Bridges Pl...             6   
14841     San Francisco Caltrain (Townsend St at 4th St)            62   
25079       Powell St BART Station (Market St at 4th St)             5   
25078       Powell St BART Station (Market St at 4th St)             5   
11420  San Francisco Ferry Building (Harry Bridges Pl...           371   

                                          station_B_name  bike_id  user_type  \
3780                             Steuart St at Market St    443.0          2   
21077                                Howard St at 2nd St   5136.0          2   
9407   San Francisco Caltrain Station 2  (Townsend St...   1814.0          2   
16632   Montgomery St BART Station (Market St at 2nd St)   4996.0          2   
6659                               Townsend St at 5th St   5015.0          3   
...                                                  ...      ...        ...   
17008                      The Embarcadero at Sansome St   5216.0          1   
14841                        Victoria Manalo Draves Park   5438.0          3   
25079       Powell St BART Station (Market St at 5th St)   4722.0          2   
25078       Powell St BART Station (Market St at 5th St)   5051.0          2   
11420                         Lombard St at Columbus Ave   5120.0          1   

       user_birth_year user_gender  
3780              1962        Male  
21077             1974        Male  
9407              1987      Female  
16632             1994      Female  
6659              1975        Male  
...                ...         ...  
17008             1966      Female  
14841             1973      Female  
25079             1994        Male  
25078             1994        Male  
11420             1969        Male  

[25698 rows x 9 columns]

### Hapus data duplicate

In [14]:
ride_sharing_new = ride_sharing_new.drop_duplicates(keep='first').reset_index(drop=True)

In [15]:
# Cek kembali apakah masih ada data duplicate 
ride_sharing_new[ride_sharing_new.duration.duplicated(keep=False)]

duration  station_A_id  \
0      12 minutes            81   
1      24 minutes             3   
2       8 minutes            67   
3       4 minutes            16   
4      11 minutes            22   
...           ...           ...   
25751  11 minutes            15   
25752  10 minutes            15   
25753  14 minutes            15   
25754  14 minutes            15   
25755  29 minutes            16   

                                          station_A_name  station_B_id  \
0                                     Berry St at 4th St           323   
1           Powell St BART Station (Market St at 4th St)           118   
2      San Francisco Caltrain Station 2  (Townsend St...            23   
3                                Steuart St at Market St            28   
4                                  Howard St at Beale St           350   
...                                                  ...           ...   
25751  San Francisco Ferry Building (Harry Bridges Pl...            34   
25752  San Francisco Ferry Building (Harry Bridges Pl...            34   
25753  San Francisco Ferry Building (Harry Bridges Pl...            42   
25754  San Francisco Ferry Building (Harry Bridges Pl...            42   
25755                            Steuart St at Market St           115   

                                      station_B_name  bike_id  user_type  \
0                                 Broadway at Kearny   5480.0          2   
1                    Eureka Valley Recreation Center   5193.0          2   
2                      The Embarcadero at Steuart St   3652.0          3   
3                       The Embarcadero at Bryant St   1883.0          1   
4                               8th St at Brannan St   4626.0          2   
...                                              ...      ...        ...   
25751                 Father Alfred E Boeddeker Park   5063.0          1   
25752                 Father Alfred E Boeddeker Park   5411.0          2   
25753  San Francisco City Hall (Polk St at Grove St)   5157.0          2   
25754  San Francisco City Hall (Polk St at Grove St)   4438.0          2   
25755                             Jackson Playground   1705.0          3   

       user_birth_year user_gender  
0                 1959        Male  
1                 1965        Male  
2                 1993        Male  
3                 1979        Male  
4                 1994        Male  
...                ...         ...  
25751             2000        Male  
25752             1998        Male  
25753             1995        Male  
25754             1995        Male  
25755             1990        Male  

[25694 rows x 9 columns]

## Case 4

### pada kolom duration terdapat satuan waktu yang dapat menghalangi dalam analisis kolom durasi pengendara sepeda. Oleh karena itu dibutuhkan cleaning data satuan waktu

In [16]:
# Remove "minutes" and convert to numeric
ride_sharing_new['duration'] = ride_sharing_new['duration'].str.replace(' minutes', '')

print(ride_sharing_new)


      duration  station_A_id  \
0           12            81   
1           24             3   
2            8            67   
3            4            16   
4           11            22   
...        ...           ...   
25751       11            15   
25752       10            15   
25753       14            15   
25754       14            15   
25755       29            16   

                                          station_A_name  station_B_id  \
0                                     Berry St at 4th St           323   
1           Powell St BART Station (Market St at 4th St)           118   
2      San Francisco Caltrain Station 2  (Townsend St...            23   
3                                Steuart St at Market St            28   
4                                  Howard St at Beale St           350   
...                                                  ...           ...   
25751  San Francisco Ferry Building (Harry Bridges Pl...            34   
25752  San Francisco Fe

## Case 5

### ubah data duration menjadi dType INT

In [17]:
ride_sharing_new['duration'] = ride_sharing_new['duration'].astype(int) # astype untuk mengubah dType menjadi integer
ride_sharing_new['duration']

0        12
1        24
2         8
3         4
4        11
         ..
25751    11
25752    10
25753    14
25754    14
25755    29
Name: duration, Length: 25756, dtype: int32

## Case 6

### Jumlah pengendara sepeda yang melewati station_A dan station_B

In [18]:
pesepeda_station_A = ride_sharing_new.groupby(['station_A_name']).size().reset_index()
pesepeda_station_A

station_A_name     0
0                                 Berry St at 4th St  3029
1                              Howard St at Beale St  2560
2   Montgomery St BART Station (Market St at 2nd St)  2997
3       Powell St BART Station (Market St at 4th St)  2899
4       Powell St BART Station (Market St at 5th St)  2386
5     San Francisco Caltrain (Townsend St at 4th St)  2932
6  San Francisco Caltrain Station 2  (Townsend St...  3635
7  San Francisco Ferry Building (Harry Bridges Pl...  2812
8                            Steuart St at Market St  2506

In [19]:
pesepeda_station_B = ride_sharing_new.groupby(['station_B_name']).size().reset_index()
pesepeda_station_B

station_B_name    0
0                                 11th St at Bryant St  183
1                                 11th St at Natoma St  184
2                                14th St at Mission St  108
3                               15th St at Potrero Ave   82
4                                 16th St Mission BART    9
..                                                 ...  ...
147                             Webster St at Grove St  105
148                         Webster St at O'Farrell St   69
149                             Williams Ave at 3rd St   12
150                          Williams Ave at Apollo St    3
151  Yerba Buena Center for the Arts (Howard St at ...  172

[152 rows x 2 columns]

### Jumlah pengendara berdasarkan user type dan gender

#### Jumlah pengendara berdasarkan user gender

In [20]:
# Filter the DataFrame Male 
pesepeda_gender_M = ride_sharing_new[ride_sharing_new['user_gender'] == 'Male']

# hitung pesepeda Male
pesepeda_male_count = pesepeda_gender_M.shape[0]

# Display
print(f"Jumlah pengendara sepeda laki-laki: {pesepeda_male_count}")

Jumlah pengendara sepeda laki-laki: 19379


In [22]:
# Filter the DataFrame Male 
pesepeda_gender_F = ride_sharing_new[ride_sharing_new['user_gender'] == 'Female']

# hitung pesepeda Male
pesepeda_female_count = pesepeda_gender_F.shape[0]

# Display
print(f"Jumlah pengendara sepeda Wanita: {pesepeda_female_count}")

Jumlah pengendara sepeda Wanita: 6026


#### Jumlah pengendara sepeda berdasar user type

In [23]:
user_type_counts = ride_sharing_new.groupby('user_type').size().reset_index()
user_type_counts

user_type      0
0          1   6285
1          2  12970
2          3   6501

## Case 7

### Berapakah rata-rata durasi pemakaian sepeda

In [24]:
rata_rata_pemakaian_sepeda = ride_sharing_new['duration'].mean()
rata_rata_pemakaian_sepeda

11.389268519956515

### Rata-rata durasi pemakaian sepeda berdasar rute A dan B

In [30]:
ride_sharing_new['duration'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 25756 entries, 0 to 25755
Series name: duration
Non-Null Count  Dtype
--------------  -----
25756 non-null  int32
dtypes: int32(1)
memory usage: 100.7 KB


In [31]:
# Convert duration to numerical value (remove ' minutes' and convert to int)
ride_sharing_new['duration'] = ride_sharing_new['duration']

# Group by 'station_A_name' and 'station_B_name' and calculate the average duration
average_duration_per_route = ride_sharing_new.groupby(['station_A_name', 'station_B_name'])['duration'].mean().reset_index(name='average_duration')

# Display the result
print(average_duration_per_route)

               station_A_name  \
0          Berry St at 4th St   
1          Berry St at 4th St   
2          Berry St at 4th St   
3          Berry St at 4th St   
4          Berry St at 4th St   
...                       ...   
1230  Steuart St at Market St   
1231  Steuart St at Market St   
1232  Steuart St at Market St   
1233  Steuart St at Market St   
1234  Steuart St at Market St   

                                         station_B_name  average_duration  
0                                  11th St at Bryant St         11.333333  
1                                  11th St at Natoma St         10.857143  
2                                 14th St at Mission St         16.500000  
3                                15th St at Potrero Ave          8.083333  
4                                  16th St Mission BART         15.333333  
...                                                 ...               ...  
1230                         Washington St at Kearny St          7.1333

### Rata-rata Usia dan Modus customer

In [36]:
# rata rata usia
rerata_usia_cust = ride_sharing_new['user_birth_year'].mean()
rerata_usia_cust

1983.055676347259

In [37]:
# rata rata usia
modus_kelahiran_cust = ride_sharing_new['user_birth_year'].std()
modus_kelahiran_cust

10.009942950728572